File: word_pattern.py
"""
290. Word Pattern

Given a pattern and a string s, find if s follows the same pattern.
Here follow means a full match, such that there is a bijection between a letter in 
pattern and a non-empty word in s.

Time Complexity: O(n + m) where n is the length of pattern and m is the length of s
Space Complexity: O(w) where w is the number of unique words
"""

class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False
        
        pattern_to_word = {}
        word_to_pattern = {}
        
        for char, word in zip(pattern, words):
            if char in pattern_to_word:
                if pattern_to_word[char] != word:
                    return False
            else:
                pattern_to_word[char] = word
            
            if word in word_to_pattern:
                if word_to_pattern[word] != char:
                    return False
            else:
                word_to_pattern[word] = char
        
        return True


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    pattern1 = "abba"
    s1 = "dog cat cat dog"
    print(f"Input: pattern = \"{pattern1}\", s = \"{s1}\"")
    print(f"Output: {solution.wordPattern(pattern1, s1)}")  # Expected: True
    print()
    
    # Example 2
    pattern2 = "abba"
    s2 = "dog cat cat fish"
    print(f"Input: pattern = \"{pattern2}\", s = \"{s2}\"")
    print(f"Output: {solution.wordPattern(pattern2, s2)}")  # Expected: False
    print()
    
    # Example 3
    pattern3 = "aaaa"
    s3 = "dog cat cat dog"
    print(f"Input: pattern = \"{pattern3}\", s = \"{s3}\"")
    print(f"Output: {solution.wordPattern(pattern3, s3)}")  # Expected: False
    print()
    
    # Additional test case
    pattern4 = "abba"
    s4 = "dog dog dog dog"
    print(f"Input: pattern = \"{pattern4}\", s = \"{s4}\"")
    print(f"Output: {solution.wordPattern(pattern4, s4)}")  # Expected: False

File: word_pattern.cpp
/*
290. Word Pattern

Given a pattern and a string s, find if s follows the same pattern.
Here follow means a full match, such that there is a bijection between a letter in 
pattern and a non-empty word in s.

Time Complexity: O(n + m) where n is the length of pattern and m is the length of s
Space Complexity: O(w) where w is the number of unique words
*/

#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <sstream>
using namespace std;

class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words = splitString(s);
        if (pattern.length() != words.size()) {
            return false;
        }
        
        unordered_map<char, string> pattern_to_word;
        unordered_map<string, char> word_to_pattern;
        
        for (int i = 0; i < pattern.length(); i++) {
            char ch = pattern[i];
            string word = words[i];
            
            if (pattern_to_word.find(ch) != pattern_to_word.end()) {
                if (pattern_to_word[ch] != word) {
                    return false;
                }
            } else {
                pattern_to_word[ch] = word;
            }
            
            if (word_to_pattern.find(word) != word_to_pattern.end()) {
                if (word_to_pattern[word] != ch) {
                    return false;
                }
            } else {
                word_to_pattern[word] = ch;
            }
        }
        
        return true;
    }
    
private:
    vector<string> splitString(const string& s) {
        vector<string> words;
        stringstream ss(s);
        string word;
        while (ss >> word) {
            words.push_back(word);
        }
        return words;
    }
};

// Test cases
int main() {
    Solution solution;
    
    // Example 1
    string pattern1 = "abba";
    string s1 = "dog cat cat dog";
    cout << "Input: pattern = \"" << pattern1 << "\", s = \"" << s1 << "\"" << endl;
    cout << "Output: " << (solution.wordPattern(pattern1, s1) ? "true" : "false") << endl;  // Expected: true
    cout << endl;
    
    // Example 2
    string pattern2 = "abba";
    string s2 = "dog cat cat fish";
    cout << "Input: pattern = \"" << pattern2 << "\", s = \"" << s2 << "\"" << endl;
    cout << "Output: " << (solution.wordPattern(pattern2, s2) ? "true" : "false") << endl;  // Expected: false
    cout << endl;
    
    // Example 3
    string pattern3 = "aaaa";
    string s3 = "dog cat cat dog";
    cout << "Input: pattern = \"" << pattern3 << "\", s = \"" << s3 << "\"" << endl;
    cout << "Output: " << (solution.wordPattern(pattern3, s3) ? "true" : "false") << endl;  // Expected: false
    cout << endl;
    
    // Additional test case
    string pattern4 = "abba";
    string s4 = "dog dog dog dog";
    cout << "Input: pattern = \"" << pattern4 << "\", s = \"" << s4 << "\"" << endl;
    cout << "Output: " << (solution.wordPattern(pattern4, s4) ? "true" : "false") << endl;  // Expected: false
    
    return 0;
}
